File: C++/239.滑动窗口最大值.cpp
/*
 * @lc app=leetcode.cn id=239 lang=cpp
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
class Solution {
public:
    deque<int> idq;
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        vector<int> res;
        for(int i = 0; i < nums.size(); i++) {
            if(i < k-1) {
                push_deque(nums[i]);
                continue;
            }
            push_deque(nums[i]);
            res.push_back(idq.front());
            if(nums[i-k+1] == idq.front()) {
                idq.pop_front();
            }
        }
        return res;
    }
    void push_deque(int val) {
        while(!idq.empty() && idq.back() < val) {
            idq.pop_back();
        }
        idq.push_back(val);
    }
};
// @lc code=end

#include <iostream>
#include <vector>

using namespace std;

int main() {
    Solution sol;

    // 测试示例 1
    vector<int> nums1 = {1, 3, -1, -3, 5, 3, 6, 7};
    int k1 = 3;
    vector<int> result1 = sol.maxSlidingWindow(nums1, k1);
    cout << "Result 1: ";
    for (int num : result1) {
        cout << num << " ";
    }
    cout << endl;

    // 测试示例 2
    vector<int> nums2 = {1, -1};
    int k2 = 1;
    vector<int> result2 = sol.maxSlidingWindow(nums2, k2);
    cout << "Result 2: ";
    for (int num : result2) {
        cout << num << " ";
    }
    cout << endl;

    // 测试示例 3
    vector<int> nums3 = {9, 11};
    int k3 = 2;
    vector<int> result3 = sol.maxSlidingWindow(nums3, k3);
    cout << "Result 3: ";
    for (int num : result3) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}

File: Go/239.滑动窗口最大值.go
/*
 * @lc app=leetcode.cn id=239 lang=golang
 *
 * [239] 滑动窗口最大值
 */

// @lc code=start
func maxSlidingWindow(nums []int, k int) []int {
	res, deque := []int{}, []int{}

	push_deque := func(val int) {
		for len(deque) > 0 && deque[len(deque)-1] < val {
			deque = deque[:len(deque)-1]
		}
		deque = append(deque, val)
	}
	for idx := 0; idx < len(nums); idx++ {
		if idx < k-1 {
			push_deque(nums[idx])
			continue
		}
		push_deque(nums[idx])
		res = append(res, deque[0])
		if deque[0] == nums[idx-k+1] {
			deque = deque[1:]
		}
	}
	return res
}

// @lc code=end

func main() {
    // 测试示例 1
    nums1 := []int{1, 3, -1, -3, 5, 3, 6, 7}
    k1 := 3
    result1 := maxSlidingWindow(nums1, k1)
    fmt.Println("Result 1:", result1)

    // 测试示例 2
    nums2 := []int{1, -1}
    k2 := 1
    result2 := maxSlidingWindow(nums2, k2)
    fmt.Println("Result 2:", result2)

    // 测试示例 3
    nums3 := []int{9, 11}
    k3 := 2
    result3 := maxSlidingWindow(nums3, k3)
    fmt.Println("Result 3:", result3)
}